cmake/gz_msgs_generate.cmake

```diff
@@ -0,0 +1,195 @@
+  set(oneValueArgs TARGET PROTO_PACKAGE MSGS_GEN_SCRIPT GZ_PROTOC_PLUGIN FACTORY_GEN_SCRIPT MSGS_LIB PROTO_PATH)
+  set(multiValueArgs INPUT_PROTOS DEPENDENCIES)
+
+  cmake_parse_arguments(generate_messages "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+  _gz_msgs_proto_pkg_to_string(${generate_messages_PROTO_PACKAGE} gen_dir)
+  _gz_msgs_proto_pkg_to_path(${generate_messages_PROTO_PACKAGE} proto_package_dir)
+
+  # Extract dependency information from targets
+  set(depends_proto_paths)
+  set(depends_includes)
+
+  file(MAKE_DIRECTORY ${PROJECT_BINARY_DIR}/${gen_dir}_gen)
+
+  foreach(dep ${generate_messages_DEPENDENCIES})
+    get_target_property(msgs_desc_file ${dep} GZ_MSGS_DESC_FILE)
+    list(APPEND depends_msgs_desc ${msgs_desc_file})
+  endforeach()
+
+  foreach(proto_file ${generate_messages_INPUT_PROTOS})
+    gz_msgs_protoc(
+      MSGS_GEN_SCRIPT
+        ${generate_messages_MSGS_GEN_SCRIPT}
+      PROTO_PACKAGE
+        ${generate_messages_PROTO_PACKAGE}
+      INPUT_PROTO
+        ${proto_file}
+      PROTOC_EXEC
+        protobuf::protoc
+      GZ_PROTOC_PLUGIN
+        ${generate_messages_GZ_PROTOC_PLUGIN}
+      PROTO_PATH
+        ${generate_messages_PROTO_PATH}
+      DEPENDENCY_PROTO_DESCS
+        ${depends_msgs_desc}
+
+      # Cpp Specific arguments
+      GENERATE_CPP
+      OUTPUT_INCLUDES
+        gen_includes
+      OUTPUT_CPP_HH_VAR
+        gen_headers
+      OUTPUT_DETAIL_CPP_HH_VAR
+        gen_detail_headers
+      OUTPUT_CPP_CC_VAR
+        gen_sources
+      OUTPUT_CPP_DIR
+        "${PROJECT_BINARY_DIR}/${gen_dir}_gen"
+
+      # Python Specific arguments
+      GENERATE_PYTHON
+      OUTPUT_PYTHON_VAR
+        gen_sources_py
+      OUTPUT_PYTHON_DIR
+        "${PROJECT_BINARY_DIR}/${gen_dir}_gen/python/"
+    )
+  endforeach()
+
+  gz_msgs_factory(
+    FACTORY_GEN_SCRIPT
+      ${generate_messages_FACTORY_GEN_SCRIPT}
+    PROTO_PACKAGE
+      ${generate_messages_PROTO_PACKAGE}
+    INPUT_PROTOS
+      ${generate_messages_INPUT_PROTOS}
+    OUTPUT_CPP_DIR
+      "${PROJECT_BINARY_DIR}/${gen_dir}_gen"
+    OUTPUT_CPP_HH_VAR
+      gen_factory_headers
+    OUTPUT_CPP_CC_VAR
+      gen_factory_sources
+    PROTO_PATH
+      ${generate_messages_PROTO_PATH}
+  )
+
+  set_source_files_properties(
+    ${gen_headers}
+    ${gen_detail_headers}
+    ${gen_sources}
+    ${gen_factory_headers}
+    ${gen_factory_sources}
+    PROPERTIES GENERATED TRUE)
+
+  if(WIN32)
+    set_source_files_properties(${gen_sources}
+      COMPILE_FLAGS "/wd4100 /wd4512 /wd4127 /wd4068 /wd4244 /wd4267 /wd4251 /wd4146")
+  endif()
+
+  if(NOT MSVC)
+    # -Wno-switch-default flags is required for suppressing a warning in some of
+    # the generated protobuf files.
+    set_source_files_properties(${gen_sources} COMPILE_FLAGS "-Wno-switch-default -Wno-float-equal")
+  endif()
+
+  set(ARGS)
+  list(APPEND ARGS -I${generate_messages_PROTO_PATH})
+  list(APPEND ARGS --descriptor_set_out="${generate_messages_TARGET}.gz_desc")
```

Comment:
This is cool. I'm guessing custom messages will now "just work" with gz-transport?

Reply:
Also, could this be used (in the future) in the message factory exclusively instead of our own message registration mechanism?

Reply:
Yes, as long as the messages are installed in the correct path, then the `gz-msgs` consumers should be able to pick them up automatically from disk.

There are some additional options to recursively include dependencies in the description, but I think that would cause issues with the one-definition bits of proto, so I haven't experimented with it.

I think there is still some benefit in keeping our factory because we have the concrete generated types, where the `gz_desc` based messages are working off of introspection.